src/engine/ult.c

```diff
@@ -610,3 +610,227 @@ dss_main_exec(void (*func)(void *), void *arg)
 
 	return dss_ult_create(func, arg, DSS_XS_SELF, info->dmi_tgt_id, 0, NULL);
 }
+
+struct dss_chore_queue {
+	d_list_t   chq_list;
+	bool       chq_stop;
+	ABT_mutex  chq_mutex;
+	ABT_cond   chq_cond;
+	ABT_thread chq_ult;
+};
+
+static void
+dss_chore_diy_internal(struct dss_chore *chore)
+{
+reenter:
+	D_DEBUG(DB_TRACE, "%p: status=%d\n", chore, chore->cho_status);
+	chore->cho_status = chore->cho_func(chore, chore->cho_status == DSS_CHORE_READY);
+	D_ASSERT(chore->cho_status != DSS_CHORE_NEW);
+	if (chore->cho_status == DSS_CHORE_READY) {
+		ABT_thread_yield();
+		goto reenter;
+	}
+}
+
+static void
+dss_chore_ult(void *arg)
+{
+	struct dss_chore *chore = arg;
+
+	dss_chore_diy_internal(chore);
+}
+
+/**
+ * Add \a chore for \a func to the chore queue of some other xstream.
+ *
+ * \param[in]	chore	address of the embedded chore object
+ * \param[in]	func	function to be executed via \a chore
+ *
+ * \retval	-DER_CANCEL	chore queue stopping
+ */
+int
+dss_chore_delegate(struct dss_chore *chore, dss_chore_func_t func)
+{
+	struct dss_module_info *info = dss_get_module_info();
+	int                     xs_id;
+	struct dss_xstream     *dx;
+	struct dss_chore_queue *queue;
+
+	chore->cho_status = DSS_CHORE_NEW;
+	chore->cho_func   = func;
+
```

Comment:
So it is the caller's duty to guarantee that it will call dss_chore_delegate() again inside func(), right? Do we have some mechanism to avoid that?

Reply:
Do you mean "it will NOT call"? If so, no, we don't check for that. If such coding bugs have occurred, we can see if it's possible to catch those bugs; for now I don't see a great need...